R/runCCA.R

```diff
@@ -157,311 +155,394 @@
+        }, character(1))
+    return(terms)
+}
+
+# This function performs dbRDA or CCA. It returns side scores with other
+# information scores in attributes.
+#' @importFrom stats as.formula na.fail
+#' @importFrom vegan cca dbrda sppscores<- eigenvals scores
+.calculate_rda <- function(
+    x, formula, data, scores, scale = TRUE, na.action = na.fail,
+     method = distance, distance = "euclidean", ord.method = "CCA", ...){
+    # input check
+    if(!.is_a_bool(scale)){
+        stop("'scale' must be TRUE or FALSE.", call. = FALSE)
+    }
+    if( !.is_a_string(method) ){
+        stop("'method' must be a single string value.", call. = FALSE)
+    }
+    if( !(.is_a_string(ord.method) && ord.method %in% c("CCA", "RDA")) ){
+      stop("'ord.method' must be 'CCA' or 'RDA'.", call. = FALSE)
     }
     #
-    # If samples do not match / there were samples without appropriate metadata
-    # and they are now removed
-    if( !all(colnames(tse) %in% rownames(rda)) && subset.result ){
-        # Take a subset
-        tse <- tse[ , rownames(rda) ]
-        # Give a message
-        warning(
-            "Certain samples are removed from the result because they did ",
-            "not include sufficient metadata.", call. = FALSE)
-    } else if( !all(colnames(tse) %in% rownames(rda)) && !subset.result ){
-        # If user do not want to subset the data
-        # Save attributes from the object
-        attr <- attributes(rda)
-        attr <- attr[ !names(attr) %in% c("dim", "dimnames")]
-        # Find samples that are removed
-        samples_not_found <- setdiff(colnames(tse), rownames(rda))
-        # Create an empty matrix
-        mat <- matrix(nrow = length(samples_not_found), ncol=ncol(rda))
-        rownames(mat) <- samples_not_found
-        # Combine the data and order it in correct order
-        rda <- rbind(rda, mat)
-        rda <- rda[colnames(tse), ]
-        # Add attributes
-        attr <- c(attributes(rda), attr)
-        attributes(rda) <- attr
+    # Get data in correct orientation. Samples should be in rows in abundance
+    # table.
+    x <- as.matrix(t(x))
+    data <- as.data.frame(data)
+    # Instead of letting na.action pass through, give informative error
+    # about missing values.
+    if( any(is.na(data)) && isTRUE(all.equal(na.action, na.fail)) ){
+        stop("Variables contain missing values. Set na.action to na.exclude",
+            " to remove samples with missing values.", call. = FALSE)
     }
-    # Add object to reducedDIm
-    reducedDim(tse, name) <- rda
-    return(tse)
+    # Ensure that the left hand side of the formula points to abundance matrix
+    formula <- as.character(formula)
+    formula[[2]] <- "x"
+    # Create a formula from string
+    formula <- as.formula(
+        paste(as.character(formula)[c(2,1,3)], collapse = " "))
+    
+    # Initialize an argument list with common arguments
+    args <- c(
+        list(formula = formula, data = data, na.action = na.action), list(...))
```

Comment:
We use always formula. If user do not want to incorporate variables, the formula is x ~ 1